controllers/PlagiarismWebhookHandler.inc.php

```diff
@@ -0,0 +1,179 @@
+
+		if (!$ithenticateSubmission) {
+			static::$_plugin->sendErrorMessage("Invalid iThenticate submission id {$payload->id} given for webhook event {$event}");
+			return;
+		}
+
+		/** @var SubmissionFileDAO $submissionFileDao */
+		$submissionFileDao = DAORegistry::getDAO('SubmissionFileDAO');
+		$submissionFile = $submissionFileDao->getById($ithenticateSubmission->submission_file_id);
+
+		if ($payload->status !== 'COMPLETE') {
+			// If the status not `COMPLETE`, then it's `ERROR`
+			static::$_plugin->sendErrorMessage(
+				"Unable to schedule the similarity report generation for file id {$submissionFile->getId()} with error : " . __("plugins.generic.plagiarism.ithenticate.submission.error.{$payload->error_code}"),
+				$submissionFile->getData('submissionId')
+			);
+			return;
+		}
+		
+		if ((int)$submissionFile->getData('ithenticateSimilarityScheduled')) {
+			static::$_plugin->sendErrorMessage("Similarity report generation process has already been scheduled for iThenticate submission id {$payload->id}", $submissionFile->getData('submissionId'));
+			return;
+		}
+		
+		list($apiUrl, $apiKey) = static::$_plugin->getServiceAccess($context);
+		$ithenticate = static::$_plugin->initIthenticate($apiUrl, $apiKey);
+
+		$scheduleSimilarityReport = $ithenticate->scheduleSimilarityReportGenerationProcess(
+			$payload->id,
+			static::$_plugin->getSimilarityConfigSettings($context)
+		);
+
+		if (!$scheduleSimilarityReport) {
+			static::$_plugin->sendErrorMessage("Failed to schedule the similarity report generation process for iThenticate submission id {$payload->id}", $submissionFile->getData('submissionId'));
+			return;
+		}
+
+		$submissionFile->setData('ithenticateSimilarityScheduled', 1);
+		$submissionFileDao->updateObject($submissionFile);
+	}
+
+	/**
+	 * Store or Update the result of similarity check for a submission file at receiving
+	 * the webook event `SIMILARITY_COMPLETE` or `SIMILARITY_UPDATED`
+	 * 
+	 * @param Context 	$context 	The current context for which the webhook request has initiated
+	 * @param string 	$payload	The incoming request payload through webhook
+	 * @param string 	$event		The incoming webhook request event
+	 *
+	 * @return void
+	 */
+	protected function storeSimilarityScore($context, $payload, $event) {
+		$payload = json_decode($payload);
+
+		// we will not store similarity check result unless it has completed
+		if ($payload->status !== 'COMPLETE') {
+			return;
+		}
+
+		$ithenticateSubmission = $this->getIthenticateSubmission($payload->submission_id);
+
+		if (!$ithenticateSubmission) {
+			static::$_plugin->sendErrorMessage("Invalid iThenticate submission id {$payload->submission_id} given for webhook event {$event}");
+			return;
+		}
+
+		/** @var SubmissionFileDAO $submissionFileDao */
+		$submissionFileDao = DAORegistry::getDAO('SubmissionFileDAO');
+		$submissionFile = $submissionFileDao->getById($ithenticateSubmission->submission_file_id);
```

Comment:
It's very unlikely that this could be abused, but it's still a good idea to at least make sure the submission file belongs to the current journal.

Reply:
Added the logic at https://github.com/pkp/plagiarism/pull/57/files#diff-47e382896266ad29f77e199ae7bbfb862e451c01ae38aa46aa5f15d0bf558fa1R211-R216
and checking at https://github.com/pkp/plagiarism/pull/57/files#diff-47e382896266ad29f77e199ae7bbfb862e451c01ae38aa46aa5f15d0bf558fa1R211-R216